service/history/api/replication/syncworkflowstate.go

```diff
@@ -0,0 +1,298 @@
+		return nil, err
+	}
+
+	newRunInfo, err := getNewRunInfo(ctx, shardContext, workflowCache, mu)
+	if err != nil {
+		return nil, err
+	}
+
+	return &historyservice.SyncWorkflowStateResponse{
+		EventBatches: events,
+		Attributes: &historyservice.SyncWorkflowStateResponse_State{
+			State: &replicationpb.SyncWorkflowStateSnapshotAttributes{
+				State: mu.CloneToProto(),
+			},
+		},
+		NewRunInfo: newRunInfo,
+	}, nil
+}
+
+func getNewRunInfo(ctx context.Context, shardContext shard.Context, workflowCache wcache.Cache, mutableState workflow.MutableState) (*replicationpb.NewRunInfo, error) {
+	newRunId := mutableState.GetExecutionInfo().GetNewExecutionRunId()
+	if len(newRunId) == 0 {
+		return nil, nil
+	}
+	mu, err := getMutableState(ctx, shardContext, namespace.ID(mutableState.GetExecutionInfo().NamespaceId), &common.WorkflowExecution{
+		WorkflowId: mutableState.GetExecutionInfo().WorkflowId,
+		RunId:      newRunId,
+	}, workflowCache)
+	if err != nil {
+		return nil, err
+	}
+	versionHistory, err := versionhistory.GetCurrentVersionHistory(mu.GetExecutionInfo().VersionHistories)
+	if err != nil {
+		return nil, err
+	}
+	eventsBlob, err := getEventsBlob(ctx, shardContext, versionHistory.BranchToken, common2.FirstEventID, common2.FirstEventID+1)
+	if err != nil {
+		return nil, err
+	}
+	if len(eventsBlob) == 0 {
+		return nil, serviceerror.NewInvalidArgument("no events found")
+	}
+	return &replicationpb.NewRunInfo{
+		RunId:      newRunId,
+		EventBatch: eventsBlob[0], // we only send first batch for new run
+	}, nil
+}
+
+func getMutableState(ctx context.Context, shardContext shard.Context, namespaceID namespace.ID, execution *common.WorkflowExecution, workflowCache wcache.Cache) (_ workflow.MutableState, retError error) {
+	wfCtx, releaseFunc, err := workflowCache.GetOrCreateWorkflowExecution(
+		ctx,
+		shardContext,
+		namespaceID,
+		execution,
+		locks.PriorityLow,
+	)
+	defer releaseFunc(retError)
+	if err != nil {
+		return nil, err
+	}
+	return wfCtx.LoadMutableState(ctx, shardContext)
+}
+
+func getEventsBlob(ctx context.Context, shardContext shard.Context, branchToken []byte, startEventId int64, endEventId int64) ([]*common.DataBlob, error) {
+	rawHistoryResponse, err := shardContext.GetExecutionManager().ReadRawHistoryBranch(ctx, &persistence.ReadHistoryBranchRequest{
+		BranchToken: branchToken,
+		MinEventID:  startEventId,
+		MaxEventID:  endEventId,
+		PageSize:    100,
+		ShardID:     shardContext.GetShardID(),
+	})
+	if err != nil {
+		return nil, err
+	}
+	return rawHistoryResponse.HistoryEventBlobs, nil
+}
+
+func getSyncStateEvents(ctx context.Context, shardContext shard.Context, targetVersionHistories *history.VersionHistories, sourceVersionHistories *history.VersionHistories, isSameBranch bool) ([]*common.DataBlob, error) {
+	startEventId := common2.EndEventID
+	sourceHistory, err := versionhistory.GetCurrentVersionHistory(sourceVersionHistories)
+	if err != nil {
+		return nil, err
+	}
+	if isSameBranch {
```

Comment:
hmm why we need this check? Can we always do `versionhistory.FindLCAVersionHistoryItemAndIndex(targetVersionHistories, sourceHistory)` ?

Reply:
Yeah, simplified. Thanks.